### PR TITLE
Added equals macro that handles null value comparison

### PR DESCRIPTION
**Pull request by adrianburusdbt**
resolves #
- https://github.com/dbt-labs/dbt-core/issues/7873
- https://github.com/dbt-labs/dbt-adapters/issues/159

### Problem
Null comparison was not handled properly in the base adapter. While some value comparisons were missing this case altogether, others were fixed but not in a repeatable, organized way.

### Solution

Based loosely on https://github.com/dbt-labs/dbt-adapters/pull/110, but ultimately it's following the suggestions made here: https://github.com/dbt-labs/dbt-core/issues/6997#issuecomment-1832992367
Extracted the logic that was in the test fixture into an 'equals.sql' macro that can be reused and replaced all other usages with this utility macro.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-adapters/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development, and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/include/global_project/macros/utils/equals.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{% macro equals(first_date, second_date, datepart) %}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/utils/equals.sql:1]: Update the params to be `expr1`, `expr2`

File: dbt/include/global_project/macros/materializations/models/incremental/merge.sql [hunk 2]
```diff
@@ -22,7 +22,7 @@
             {% endfor %}
         {% else %}
             {% set unique_key_match %}
-                DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
+                {{ adapter.dispatch('equals', 'dbt')(DBT_INTERNAL_SOURCE.{{ unique_key }}, DBT_INTERNAL_DEST.{{ unique_key }}) }}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/materializations/models/incremental/merge.sql:25]: `adapter.dispatch('equals', 'dbt')` should just be `equals`. The dispatch happens within the `equals` macro, which then dispatches to `default__equals` or the platform-specific version, e.g. `postgres__equals`.

I would also test this against a concrete adapter, like `dbt-postgres`. I don't know if you can do the double expansion (`{{}}` inside another `{{}}`).

File: dbt/include/global_project/macros/materializations/models/incremental/merge.sql [hunk 3]
```diff
@@ -62,11 +62,11 @@
 
     {% if unique_key %}
         {% if unique_key is sequence and unique_key is not string %}
-            delete from {{target }}
+            delete from {{ target }}
             using {{ source }}
             where (
                 {% for key in unique_key %}
-                    {{ source }}.{{ key }} = {{ target }}.{{ key }}
+                    {{ adapter.dispatch('equals', 'dbt')({{ source }}.{{ key }}, {{ target }}.{{ key }}) }}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/materializations/models/incremental/merge.sql:69]: Same here, just `equals`.

File: dbt/include/global_project/macros/materializations/snapshots/helpers.sql [hunk 4]
```diff
@@ -53,8 +53,7 @@
         from {{ target_relation }}
         where
             {% if config.get('dbt_valid_to_current') %}
-               {# Check for either dbt_valid_to_current OR null, in order to correctly update records with nulls #}
-               ( {{ columns.dbt_valid_to }} = {{ config.get('dbt_valid_to_current') }} or {{ columns.dbt_valid_to }} is null)
+               {{ adapter.dispatch('equals', 'dbt')({{ columns.dbt_valid_to }}, {{ config.get('dbt_valid_to_current') }}) }}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/materializations/snapshots/helpers.sql:56]: Same here, just `equals`.

File: dbt/include/global_project/macros/materializations/snapshots/helpers.sql [hunk 5]
```diff
@@ -276,7 +275,7 @@
 {% macro unique_key_join_on(unique_key, identifier, from_identifier) %}
     {% if unique_key | is_list %}
         {% for key in unique_key %}
-            {{ identifier }}.dbt_unique_key_{{ loop.index }} = {{ from_identifier }}.dbt_unique_key_{{ loop.index }}
+            {{ adapter.dispatch('equals', 'dbt')({{ identifier }}.dbt_unique_key_{{ loop.index }}, {{ from_identifier }}.dbt_unique_key_{{ loop.index }}) }}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/materializations/snapshots/helpers.sql:278]: Same here, just `equals`.

File: dbt/include/global_project/macros/materializations/snapshots/snapshot_merge.sql [hunk 6]
```diff
@@ -15,8 +15,8 @@
 
     when matched
      {% if config.get("dbt_valid_to_current") %}
-       and (DBT_INTERNAL_DEST.{{ columns.dbt_valid_to }} = {{ config.get('dbt_valid_to_current') }} or
-            DBT_INTERNAL_DEST.{{ columns.dbt_valid_to }} is null)
+       and {{ adapter.dispatch('equals', 'dbt')(DBT_INTERNAL_DEST.{{ columns.dbt_valid_to }}, {{ config.get('dbt_valid_to_current') }}) }}
```
Comment (mikealfare) [inline dbt/include/global_project/macros/materializations/snapshots/snapshot_merge.sql:18]: Same here, just `equals`.

**Review by VersusFacit — APPROVED** [review]
thanks @mikealfare for doing the bulk of CR here.

**Pull request merged**